### PR TITLE
Add useToggle hook for delay open of tooltip

### PR DESCRIPTION
**Pull request by rwharrisjr**
Adds `useToggle` as a hook for controlling components with an "open" prop. Provides

- isOpen state prop
- function to open by event
- function to close by event
- function to toggle by event

`const {isOpen, open, close} = useToggle(400, 0);`

### CODE REVIEW

**Comment by mvelkCisco** [comment]
I agree `useEffect` delaying internal state change is nasty. This is more neat and basically what we do in https://github.com/advthreat/intelligence/blob/master/src/hooks/useTooltip.js. 

However it doesn't solve the original issue I had. My intention was to also delay this `infoTooltip` that's build in `AFieldBase` https://github.com/cisco-sbg-ui/magna-react/blob/canary/framework/components/AFieldBase/AFieldBase.js#L63.
Maybe we can use `useToggle` in `AFieldBase` and expose its args as a `AFieldBase` prop?

File: framework/components/ATooltip/ATooltip.mdx [hunk 1]
```diff
@@ -16,7 +16,33 @@ import {ATooltip} from "@cisco-sbg-ui/magna-react";
 
 It is necessary for `ATooltip` to be a descendant of `AApp` for mounting.
 
-## Usage
+## Usage (hook with delay options)
+
+<Playground
+  code={`() => {
+  const iconRef = useRef(null);
+  const {isOpen, open, close} = useToggle(400, 0);
```
Comment (mvelkCisco) [inline framework/components/ATooltip/ATooltip.mdx:24]: For the example I would loose the args of `useToggle` to encorage using defaults for consistency.

```suggestion
  const {isOpen, open, close} = useToggle();
```

File: framework/hooks/useToggle.js [hunk 2]
```diff
@@ -0,0 +1,36 @@
+import {useCallback, useRef, useState} from "react";
+
+const useToggle = (openDelay = 400, closeDelay = 0) => {
```
Comment (mvelkCisco) [inline framework/hooks/useToggle.js:3]: `closeDelay` should default to `undefined` so that `openDelay / 2` fallback is used.

**Comment by mvelkCisco** [comment]
Speaking of the tooltip, maybe including an example on how to write a wrapper for `ATooltip` (something like this https://github.com/advthreat/intelligence/blob/master/src/components/Tooltip.js#L41) would make the life of new comers and maybe also existing users easier. Or maybe just straight up export it as `AHoverTooltip` or something like that.

Don't get me wrong, I love the flexibility you are aiming for. But adding `useRef`, `useState` (now `useToggle`) and 2 event handlers for the tooltip target every single time is really not ergonomic (and eg in data table cell I would say not advised).

In 99% of cases I just want the tooltip to show and hide on hover, with a consistent delay. That can be as easy as `<AHoverTooltip content="foo">bar</AHoverTooltip>` with a wrapper component.

PS: My wrapper has some downsides, it wraps multiple children/number/string with a `<span>` to be able to attach `ref` and does not work with react function components that don't implement `forwardRef` (that can be fixed with an extra `<span>` too). In some case that `span` might cause css issues.

**Comment by rwharrisjr** [comment]
I added `useTooltip` to `AFieldBase`. I think it's safe enough to make that standard behavior in this situation. 


I 100% agree with making `AHoverTooltip`.  The requirement for components using `forwardRef` in your implementation is pretty standard across tooltip libraries I've seen (@tippyjs/react). I will add a similar implementation in magna-react.

**Comment by rwharrisjr** [comment]
@mvelkCisco Added `ATriggerTooltip` to this PR

Usage will be 

```
  <ATriggerTooltip
    id="tooltip_usage_1"
    placement="bottom-left"
    content="Tooltip text"
    pointer>
     <AIcon>
    information</AIcon>
  </ATooltip>
```

File: framework/hooks/useToggle.js [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+      closeDelay ?? openDelay / 2
+    );
+  }, [closeDelay, openDelay, timeout]);
+
+  const toggle = useCallback(() => {
+    isOpen ? close() : open();
+  }, [isOpen, open, close]);
+
+  useEffect(() => {
+    timeout.current && clearTimeout(timeout.current);
```
Comment (brennarvo) [inline framework/hooks/useToggle.js:34]: Can you describe what this `useEffect` is for? It seems like whenever we change `timeout` in `useToggle`, we clear the timeout anyways, such as in the `open` and `close` functions.

If it's for a cleanup, it needs to be a returned function from `useEffect`

Comment (rwharrisjr) [reply framework/hooks/useToggle.js:34]: That is meant to be cleanup. My brain skipped the needing to return it when I added that.

File: framework/components/AFieldBase/AFieldBase.js [hunk 4]
```diff
@@ -52,25 +50,15 @@ const AFieldBase = forwardRef(
...
-                  {...infoTooltipProps}
-                >
-                  {infoTooltip}
-                </ATooltip>
-              </>
+              <ATriggerTooltip
+                id={`${labelId}-tooltip`}
+                placement="top"
+                content={infoTooltip}
+                pointer
```
Comment (mvelkCisco) [inline framework/components/AFieldBase/AFieldBase.js:57]: Can you please make `pointer` `true` a default in `ATooltip`? I don't think I've ever seen a tooltip without a pointer in magnetic. This is no-pointer version is coming from atomic.

Comment (rwharrisjr) [reply framework/components/AFieldBase/AFieldBase.js:57]: Done

File: framework/components/ATriggerTooltip/ATriggerTooltip.js [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+    );
+  });
+
+  return (
+    <>
+      {React.Children.map(children, (child, index) =>
+        React.cloneElement(child, {
+          ...child.props,
+          ref: (ref) => {
+            childrenRef.current[index] = ref;
```
Comment (brennarvo) [inline framework/components/ATriggerTooltip/ATriggerTooltip.js:55]: We'll also want to pass the `ref` passed to this internal function to a possible `forwardRef` to the `ATriggerTooltip` children.

Like:

```jsx
function MyFeature() {
  const featureRef = useRef();

  return (
    <ATriggerTooltip>
      <p ref={featureRef}>I am the trigger for the tooltip component I am wrapped in.</p>
    </ATriggerTooltip>
  )
}
```
We'll want the `MyFeature` component to be able to access the `ref` on the `<p>` element.

```diff
ref: (node) {
  childrenRef.current[index] = ref;
+  if (typeof triggerTooltipForwardedRef === 'function') {
+  } else {
+   triggerTooltipForwardedRef.current = node;
+  }
}
```

File: framework/components/ATriggerTooltip/ATriggerTooltip.js [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+  const childrenRef = useRef([]);
+  const {isOpen, open, close, toggle} = useToggle(openDelay, closeDelay);
+
+  useEffect(() => {
+    const childRefs = childrenRef.current;
+    switch (trigger) {
+      case "mouseenter":
+        childRefs.forEach((childRef) => {
+          childRef.addEventListener("mouseenter", open);
+          childRef.addEventListener("mouseleave", close);
```
Comment (brennarvo) [inline framework/components/ATriggerTooltip/ATriggerTooltip.js:24]: Is there a reason for not adding these event listeners where we clone the child, i.e.,:

```js
return (
  <>
    {React.Children.map(children, (child, index) =>
      React.cloneElement(child, {
        ...child.props,
        onMouseEnter: (e) => {
          open();
          const childOnMouseEnter = child.props.onMouseEnter;
          childOnMouseEnter && childOnMouseEnter(e);
        },
        onMouseLeave: (e) => {
          close();
          const childOnMouseLeave = child.props.onMouseLeave;
          childOnMouseLeave && childOnMouseLeave(e);
        },
        onClick: (e) => {
          open();
          const childOnClick = child.props.onClick;
          childOnClick && childOnClick(e);
        },
        ref: (ref) => {
          childrenRef.current[index] = ref;
        },
      })
    )}
    {childElements}
  </>
)
```

File: framework/components/ATriggerTooltip/ATriggerTooltip.js (hunk 5 above)
Comment (brennarvo) [reply framework/components/ATriggerTooltip/ATriggerTooltip.js:87]: Or, if you want to use the utility in Magna React:

```diff
+ ref: handleMultipleRefs(props.ref, (node) => {
+   childrenRef.current[index] = node;
+ })
```

**Review by brennarvo — APPROVED** [review]
👍

**Pull request merged**